### PR TITLE
Add Ruma API Coverage Final GSoC blog post

### PR DESCRIPTION
**Pull request by Frinksy**
Here's a first draft, let me know what you think!
Some parts might be too wordy, but I'd be happy to condense those if it would be more appropriate.

### CODE REVIEW

File: content/news/2021-08-19-gsoc-coverage.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+sharing part was possible to implement with the current macros available in
+Ruma, storage requires a major refactoring of these macros. This means that
+although I was able to do the groundwork for supporting [MSC1946], secret
+storage will not be usable until those macros are reworked. 
+
+The reason for all this trouble is that [MSC1946] introduces an event
+`*m.secret_storage.profile.[key_ID]*` where `[key_ID]` is a placeholder for the
+ID of a key to be stored. This means there could be any number of event types
+possible, instead of the until now predefined set of event types. I thus opened
+a [PR] for this to be supported in Ruma.
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: It's not a PR, it's an issue ;)

File: content/news/2021-08-19-gsoc-coverage.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ID of a key to be stored. This means there could be any number of event types
+possible, instead of the until now predefined set of event types. I thus opened
+a [PR] for this to be supported in Ruma. 
+
+Although secret storage will need to wait to be merged, secret sharing has
+already been merged. This part of [MSC1946], as its name suggests, allows
+clients to share secrets between devices. As such, keys can be exchanged between
+devices using the  part of the Matrix Spec introduced by this MSC. In fact,
+[`matrix-rust-sdk`] has already implemented [the gossiping of cross signing
+keys], building on top of Ruma's implementation of Secret Sharing.
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: Missing link for gossiping of cross signing keys.

File: content/news/2021-08-19-gsoc-coverage.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+As I wasn't realistic to include the major macro refactor in my GSoC project, I
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]:
```suggestion
As it wasn't realistic to include the major macro refactor in my GSoC project, I
```

File: content/news/2021-08-19-gsoc-coverage.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+decided to move onto  [MSC2785: Event notification attributes and actions]. I
+had kept this as an extra goal for my project, and seemed to be a good fit for
+the situation. Unluckily though, when I read more into the MSC's text, I
+realized I would encounter the exact same problem as I had with Secret Storage.
+This MSC introduced the event type `*m.notifications_profile.<profile>*` where
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: I think you should decide for either emphasis (`*`s) or an inline code block, not use both.

File: content/news/2021-08-19-gsoc-coverage.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+all the currently laid out endpoints and events have been implemented in a
+branch ready to be merged once possible.
+
+Thus there are a couple of pull requests which are blocked and that will be
+merged in the future, once variable event type names are supported:
+
+- [Add support for MSC2785: Event notification attributes and actions]
+- [Add secret storage]
+
+Finally, I worked on a final issue for my GSoC project:
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: 2x "final" in one short sentence

File: content/news/2021-08-19-gsoc-coverage.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+- [Generate for impls for event enums]
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]:
```suggestion
- [Generate From impls for event enums]
```

File: content/news/2021-08-19-gsoc-coverage.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+To solve this last issue, I decided to learn how to write procedural macros. In
+the first half of GSoC, I had created a new `macro_rules!` macro, largely based
+on a pre-existing one. As I had very little experience in writing macros, and
+zero experience in writing procedural ones, I decided to read a short series of
+[blog posts] to get myself up to speed. Once that was done, I got to work: I
+added the generation of `From` trait implementations inside the pre-existing
+`[proc-macro]` functions which generate the event enums. I also added a `Derive`
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: Maybe don't highlight `[proc-macro]` like it was some sort of syntax within Rust and just have sth. like

```suggestion
proc-macro code which generate the event enums. I also added a `Derive`
```

Comment (jplatte) [reply content/news/2021-08-19-gsoc-coverage.md]: Same for "Derive", really. I don't think it needs to be captialized or highlighted.

File: content/news/2021-08-19-gsoc-coverage.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+
+
+[previous blog post]: https://www.ruma.io/news/gsoc-2021-intro/
+[3PID validity endpoint]: https://github.com/ruma/ruma/pull/618
+[3PID bind endpoint]: https://github.com/ruma/ruma/pull/621
+[Invitation signing endpoint]: https://github.com/ruma/ruma/pull/626
+[Store invitation endpoint]: https://github.com/ruma/ruma/pull/631
+[3PID unbind endpoint]: https://github.com/ruma/ruma/pull/627
+[Add client secret and session identifier types]: eat
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: What happened here? 😄

Comment (Frinksy) [reply content/news/2021-08-19-gsoc-coverage.md]: I think I had clicked "copy" instead of "copy link", and had "eat" selected randomly somewhere. I guess this one slipped past :sweat_smile:

File: content/news/2021-08-19-gsoc-coverage.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
+# Ruma API Coverage GSoC Coming to an End
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: Please add TOML frontmatter with a `title` (removing this heading) and `author = "Adam Blanchet"`. TOML frontmatter is enclosed in two lines that are just three `+` signs. You can check the other files in this directory and/or install [Zola](https://www.getzola.org/documentation/getting-started/overview/) and run `zola serve` locally to check out what it looks like.

I think I'd also like to kinda sync the titles of the two end-of-gsoc blog posts, let's discuss that on Matrix.

File: content/news/2021-08-19-gsoc-coverage.md [hunk 10]
```diff
@@ -1,4 +1,8 @@
-# Ruma API Coverage GSoC Coming to an End
++++
+title = "GSoC 2021: API Coverage Final Summary"
+date = 2021-08-22
```
Comment (jplatte) [inline content/news/2021-08-19-gsoc-coverage.md]: Please adjust the file name accordingly, then this can be removed.
